src/main/java/com/puppycrawl/tools/checkstyle/xpath/ElementNode.java

```diff
@@ -0,0 +1,234 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.xpath;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import net.sf.saxon.om.AxisInfo;
+import net.sf.saxon.om.NodeInfo;
+import net.sf.saxon.tree.iter.ArrayIterator;
+import net.sf.saxon.tree.iter.AxisIterator;
+import net.sf.saxon.tree.iter.EmptyIterator;
+import net.sf.saxon.tree.iter.SingleNodeIterator;
+import net.sf.saxon.tree.util.Navigator;
+import net.sf.saxon.type.Type;
+
+/**
+ * Represents element node of Xpath-tree.
+ *
+ * @author Timur Tibeyev
+ */
+public class ElementNode extends AbstractNode {
+    /** String literal for text attribute. */
+    private static final String TEXT_ATTRIBUTE_NAME = "text";
+
+    /** The parent of the current node. */
+    private final AbstractNode parent;
+
+    /** The root node. */
+    private final AbstractNode root;
+
+    /** The ast node. */
+    private final DetailAST detailAst;
+
+    /** Represents value of TokenTypes#IDENT. */
+    private String ident;
+
+    /** Represents text of the DetailAST. */
+    private String text;
+
+    /** The attributes. */
+    private AbstractNode[] attributes;
```

Comment:
done, `detalAst` is `final` thats why current order is root, parent, detalAst, attributes, ident, text